components/script/dom/htmlimageelement.rs

```diff
@@ -866,14 +926,6 @@ impl HTMLImageElement {
         let src = elem.get_url_attribute(&local_name!("src"));
         let base_url = document.base_url();
 
-        // https://html.spec.whatwg.org/multipage/#reacting-to-dom-mutations
```

Comment:
I propose that we resurrect this block, but we change pending_request to Unavailable instead.

Reply:
Based on the latest WPT run, looks like we're back to to the two `<img>.complete` failures.  I'm guessing this suggestion aimed to resolve those failures, so I gave it a go.

---

**Attempt one:**
_Only setting `State::Unavailable` if `pending_request` is `Some`:_
```rust
{
    if let Some(ref mut pending_request) = *self.pending_request.borrow_mut() {
        pending_request.state = State::Unavailable;
    }
}
```

**Results:**
_Excludes the `assert_unreached: ref currentSrc was "" Reached unreachable code` you resolved separately_
```
Ran 45 tests finished in 101.0 seconds.
  • 38 ran as expected. 0 tests skipped.
  • 7 tests had unexpected subtest results

Tests with unexpected results:
  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async pending request test
  │   → assert_false: Should not be complete because we have started a new load expected false got true
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:197:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/update-src-complete.html:
  │ FAIL [expected PASS] Changing the img src should retain the 'complete' property
  │   → uncaught exception: unknown (can't convert to string)
  │ 
  │ check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:12:9
  └ onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:20:9

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: source (max-width:500px) broken image, img valid image, resize to wide

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: source (max-width:500px) valid image, img valid image, resize to wide

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: source (max-width:500px) valid image, img broken image, resize to narrow

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: source (max-width:500px) valid image, img valid image, resize to narrow
```

---

**Attempt two:**
_Unconditionally setting `pending_request` to `State::Unavailable`, initializing `pending_request` to `ImageRequest::new()` if it is `None`:_

```rust
{
    let mut pending_opt = self.pending_request.borrow_mut();
    pending_opt.get_or_insert(ImageRequest::new()).state = State::Unavailable;
}
```

**Results:**

```
Running 45 tests in web-platform-tests

  /html/semantics/embedded-content/the-img-element/img-with-containment-and-size.html
  /html/semantics/embedded-content/the-img-element/document-base-url.html
  /html/semantics/embedded-content/the-img-element/document-adopt-base-url.html
  /html/semantics/embedded-content/the-img-element/available-images.html
  /html/semantics/embedded-content/the-img-element/sizes/sizes-dynamic-002.html
  /html/semantics/embedded-content/the-img-element/sizes/sizes-dynamic-001.html
  /html/semantics/embedded-content/the-img-element/404-response-with-actual-image-data.html
  /html/semantics/embedded-content/the-img-element/decode/image-decode-path-changes.html
  /html/semantics/embedded-content/the-img-element/Image-constructor.html
  /html/semantics/embedded-content/the-img-element/adoption.html
  /html/semantics/embedded-content/the-img-element/current-pixel-density/error.html
  /html/semantics/embedded-content/the-img-element/current-pixel-density/basic.html
  /html/semantics/embedded-content/the-img-element/null-image-source.html
  /html/semantics/embedded-content/the-img-element/data-url.html
  /html/semantics/embedded-content/the-img-element/decode/image-decode-picture.html
  /html/semantics/embedded-content/the-img-element/historical-progress-event.window.html
  /html/semantics/embedded-content/the-img-element/decode/image-decode-svg.tentative.html
  /html/semantics/embedded-content/the-img-element/available-images-onload.html
  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async srcset complete test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:108:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async src removal test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:161:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async src broken test
  │   → assert_true: expected true got false
  │ 
  │ img.onerror<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:151:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async pending request test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:195:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async srcset removal test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:173:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async src complete test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:89:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  /html/semantics/embedded-content/the-img-element/decode/image-decode.html
  /html/semantics/embedded-content/the-img-element/decode/image-decode-with-quick-attach.html
  /html/semantics/embedded-content/the-img-element/invalid-src.html
  /html/semantics/embedded-content/the-img-element/nonexistent-image.html
  /html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-quirks-mode.html
  /html/semantics/embedded-content/the-img-element/delay-load-event-detached.html
  /html/semantics/embedded-content/the-img-element/relevant-mutations.html
  /html/semantics/embedded-content/the-img-element/delay-load-event.html
  /html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-width-1000px.html
  /html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-display-none.html
  /html/semantics/embedded-content/the-img-element/srcset/select-an-image-source.html
  /html/semantics/embedded-content/the-img-element/srcset/parse-a-srcset-attribute.html
  /html/semantics/embedded-content/the-img-element/non-active-document.html
  /html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-standards-mode.html
  /html/semantics/embedded-content/the-img-element/srcset/srcset-media-dynamic.html
  /html/semantics/embedded-content/the-img-element/usemap-casing.html
  /html/semantics/embedded-content/the-img-element/srcset/avoid-reload-on-resize.html
  /html/semantics/embedded-content/the-img-element/not-rendered-dimension-getter.html
  /html/semantics/embedded-content/the-img-element/update-the-image-data/fail-to-resolve.html
  /html/semantics/embedded-content/the-img-element/update-media.html
  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/update-src-complete.html:
  │ FAIL [expected PASS] Changing the img src should retain the 'complete' property
  │   → uncaught exception: unknown (can't convert to string)
  │ 
  │ check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:10:9
  └ onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:20:9

  /html/semantics/embedded-content/the-img-element/update-the-source-set.html
  /html/semantics/embedded-content/the-img-element/decode/image-decode-iframe.html
  /html/semantics/embedded-content/the-img-element/decode/image-decode-path-changes-svg.tentative.html
  /html/semantics/embedded-content/the-img-element/decode/image-decode-with-quick-attach-svg.tentative.html
  ▶ TIMEOUT [expected OK] /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html
  │ 
  │ 
  │ 
  └ 

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] img (srcset 1 cand) valid image, resize to wide

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) broken image, img valid image, resize to wide
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) valid image, img valid image, resize to wide
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: same URL in source (max-width:500px) and img, resize to wide

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] img (srcset 1 cand) valid image, resize to narrow

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) valid image, img broken image, resize to narrow
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) valid image, img valid image, resize to narrow
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: same URL in source (max-width:500px) and img, resize to narrow


Ran 45 tests finished in 103.0 seconds.
  • 42 ran as expected. 0 tests skipped.
  • 1 tests timed out unexpectedly
  • 3 tests had unexpected subtest results

Tests with unexpected results:
  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async srcset complete test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:108:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async src removal test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:161:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async src broken test
  │   → assert_true: expected true got false
  │ 
  │ img.onerror<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:151:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async pending request test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:195:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async srcset removal test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:173:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async src complete test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:89:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/update-src-complete.html:
  │ FAIL [expected PASS] Changing the img src should retain the 'complete' property
  │   → uncaught exception: unknown (can't convert to string)
  │ 
  │ check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:10:9
  └ onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:20:9

  ▶ TIMEOUT [expected OK] /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html
  │ 
  │ 
  │ 
  └ 

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] img (srcset 1 cand) valid image, resize to wide

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) broken image, img valid image, resize to wide
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) valid image, img valid image, resize to wide
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: same URL in source (max-width:500px) and img, resize to wide

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] img (srcset 1 cand) valid image, resize to narrow

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) valid image, img broken image, resize to narrow
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) valid image, img valid image, resize to narrow
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: same URL in source (max-width:500px) and img, resize to narrow
```

These results are worse, failing on the first `<img>.complete` in tests are where there are two `<img>.complete` calls.

---

Are either of these what you were thinking?  I've pushed the code for attempt one in case that helps illustrate the situation.  I can try to dig deeper later to see if there's some way we can make this work and figure out why these code changes have these results, but wanted to update with my initial findings.